allure-multi/action.py

```diff
@@ -0,0 +1,161 @@
+        ],
+        check=True,
+        capture_output=True,
+    )
+    return json.loads(runs.stdout)
+
+
+def download_run(repository, run_id, pattern=PATTERN):
+    """
+    Download artifacts for a single run.
+    """
+    assert pattern, "pattern is required"
+    workdir = Path(run_id)
+    try:
+        workdir.mkdir()
+    except OSError:
+        print(f"Skip existing {workdir}")
+        return workdir
+    result = subprocess.run(
+        ["gh", "run", "download", "-p", pattern, "-R", repository, run_id],
+        cwd=workdir,
+    )
+    if result.returncode == 0:
+        return workdir
+    else:
+        # will also fail if workdir is not empty
+        workdir.rmdir()
+        print("stdout", result.stdout)
+        print("stderr", result.stderr)
+
+
+def report_run(
+    repository: str,
+    run: dict,
+    outdir="gh-pages",
+    pattern: re.Pattern = re.compile(r"(.*)"),
+):
+    workdir = Path(str(run["databaseId"]))
+    groups = {}
+    for path in workdir.iterdir():
+        if match := pattern.match(path.name):
+            groupname = match.groups()[0]
+            groups[groupname] = groups.get(groupname, []) + [path]
+
+    for group in groups:
+        report_dir = Path(workdir, f"allure-report-{group}")
+        report_dir.mkdir(exist_ok=True)
+        # extract related results on top of each other
+        for artifact_dir in groups[group]:
+            command = [
+                "find",
+                str(artifact_dir),
+                "-name",
+                "allure-results.tar.gz",
+                "-exec",
+                "tar",
+                "-C",
+                str(report_dir),
+                "-xf",
+                "{}",
+                "--strip-components=1",
+                ";",  # no shell escaping
+            ]
+
+            subprocess.run(
+                command,
+            )
+
+        output_dir = Path(outdir, run["headBranch"].replace("/", "-"), group)
+
+        try:
+            os.rename(output_dir / "history", report_dir / "history")
+        except OSError:
+            print(f"Could not move history from {output_dir} to {report_dir}")
+
+        # allure generate --output <report dir> input-dir
+        command = [
+            ALLURE,
+            "generate",
+            "--clean",
+            "--output",
+            str(output_dir),
+            report_dir,
+        ]
+        subprocess.run(command)
+
+
+def index_html(outdir="gh-pages"):
+    """
+    Link to all the reports.
+    """
+    outdir = Path(outdir)
+    outfile = outdir / "index.html"
+    indexes = [
+        index.relative_to(outdir) for index in Path(outdir).glob("*/**/index.html")
+    ]
+
+    page = """\
+<link rel='stylesheet' href='https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/css/bootstrap.min.css'>
+<div class='container'>
+<h1>Test reports</h1>
+""" + "\n".join(
+        (f"""<a href="{index}">{index}</a><br>""" for index in indexes)
+    )
```

Comment:
Let's support code folding, without this, VSCode and all other editors will do:

<img width="782" alt="Screen Shot 2022-11-22 at 11 15 23 AM" src="https://user-images.githubusercontent.com/4546435/203380753-89cfe9f5-562a-47b2-9b0e-4c1888f78f6a.png">

```suggestion
    page = textwrap.dedent(
        """\
        <link rel='stylesheet' href='https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/css/bootstrap.min.css'>
        <div class='container'>
            <h1>Test reports</h1>
        """
    ) + "\n".join(
        (f"""<a href="{index}">{index}</a><br>""" for index in indexes)
    )
```

Reply:
The correct code folding for the unindented code is only possible in VSCode if the Python extension is installed (no guarantee for other editors).

Reply:
This is interesting, to be able to glance over 3 lines of code more quickly. It would also be possible to include extra whitespace in indentation-insensitive HTML.